### ISSUE TITLE
[sklearn#34387] `OrdinalEncoder` surprising behavior: missing values are treated as unknown categories first

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#328]
**Mirrored from [scikit-learn#34387](https://github.com/scikit-learn/scikit-learn/issues/34387)**

Original title: `OrdinalEncoder` surprising behavior: missing values are treated as unknown categories first
Original labels: Bug, Needs Triage
Created: 2026-06-25T09:55:30Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the bug and give evidence about its user-facing impact

I find many behaviors around missing values and it interacts with `handle_unknown` and `unknown_value` quite surprising (and probably quite not what you want). It's surprising because to me, nothing in the code lets you anticipate that `handle_unknown` and `unknown_value` might impact of missing values are handle.

Below a list of examples with comments at the format `# actual behavior | desirable behavior (IMO)`

### Steps/Code to Reproduce

```python
import numpy as np
from sklearn.preprocessing import OrdinalEncoder

encoder = OrdinalEncoder(encoded_missing_value=np.nan, unknown_value=-1, handle_unknown='use_encoded_value')
# returns nan | returns nan
encoder.fit([[0], [1], [np.nan]]).transform([[np.nan]])
# returns -1  | returns nan
encoder.fit([[0], [1]]).transform([[np.nan]])

encoder = OrdinalEncoder(encoded_missing_value=np.nan)
# returns nan | returns nan
encoder.fit([[0], [1], [np.nan]]).transform([[np.nan]])
# raises | returns nan
encoder.fit([[0], [1]]).transform([[np.nan]])
# raises | returns nan 
encoder.fit(np.array([[0], [1], [np.nan]], dtype=object)).transform([[np.nan]])
# ^ this one is definitely a bug, not just a surprising behavior
# it can be fixed by a small local change, contrary to 

encoder = OrdinalEncoder(encoded_missing_value=np.nan, unknown_value=np.nan, handle_unknown='use_encoded_value')
# raises | returns nan
encoder.fit(np.array([[0], [1]], dtype=object)).transform(np.array([[np.nan]]))
# ^ same bug as a above
```

### Expected/Actual Results

Mentioned in comments with format `# actual behavior | desirable behavior (IMO)`

### Versions

```shell
System:
    python: 3.14.3 (main, Feb 12 2026, 00:42:54) [Clang 21.1.4 ]
executable: /home/arthur/open-source/skl/scikit-learn/sklearn-env/bin/python
   machine: Linux-6.17.0-35-generic-x86_64-with-glibc2.39

Python dependencies:
      sklearn: 1.10.dev0
          pip: None
   setuptools: 70.2.0
        numpy: 2.4.6
        scipy: 1.17.1
       Cython: 3.2.5
       pandas: 3.0.3
   matplotlib: 3.10.9
       joblib: 1.5.3
threadpoolctl: 3.6.0
     narwhals: 2.21.2

Built with OpenMP: True

threadpoolctl info:
       user_api: blas
   internal_api: openblas
    num_threads: 16
...
    num_threads: 16
         prefix: libgomp
       filepath: /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0
        version: None
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...
```

### Interest in fixing the bu

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/95 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
